### ISSUE TITLE
Stream result after cast evaluated as having @PolyNull element

### ISSUE THREAD

**Issue by MichalStehlikCz** [typetools/checker-framework#6739]
Compilation of

public class Test {

  public <T> Collection<T> clear(Class<T> type) {
    return Stream.of("abc")
        .map(type::cast)
        .toList();
  }
}

fails with

java: [return] incompatible types in return.
  type of expression: @Initialized @NonNull List<T extends @Initialized @PolyNull Object>
  method return type: @Initialized @NonNull Collection<T extends @Initialized @Nullable Object>

function with @PolyNull annotated return type should evaluate based on its argument type (e.g. @NonNull in this case)

Worked fine in 3.42.0, broken in 3.46.0

**Comment by MichalStehlikCz** [comment]
... works fine when lambda is used instead of method reference

```
public class Test {

  public <T> Collection<@NonNull T> clear(Class<T> type) {
    return Stream.of("abc")
        .map(value -> type.cast(value))
        .toList();
  }
}

```

**Comment by mernst** [comment]
@MichalStehlikCz Thanks for the bug reports; we appreciate it.

**Comment by DavyLandman** [comment]
I've made a small repo case. And there's even something weirder going one, since if I remove `CustomStream` the error meessage for the `x.stream` case flips from `PolyNull` to `Nullable`. Also on 3.50, going back to 3.42 solves it.

```java
    private static class CustomStream {
        private final Object[] data;

        CustomStream(Object[] data) {
            this.data = data;
        }

        Stream<Object> stream() {
            return Arrays.stream(data);
        }
        
    }


    public static void main(String[] args) {
        List<Object> x = List.of("abc", "bbc2");

        int total = 0;
        for (var e : x) {
            total += String.class.cast(e).length();
        }
        System.out.println(total);

        total = x.stream()
            .map(String.class::cast)
            .mapToInt(String::length)
            .sum();

        System.out.println(total);

        var custom = new CustomStream(x.toArray());

        total = custom.stream()
            .map(String.class::cast)
            .mapToInt(String::length)
            .sum();

        System.out.println(total);

    }
```
